src/pyeval.jl

```diff
@@ -2,29 +2,38 @@ const Py_single_input = 256  # from Python.h
 const Py_file_input = 257
 const Py_eval_input = 258
 
-const _maindict = PyDict{String,PyObject,false}(PyNULL()) # cache of __main__ module dictionary
-function maindict()
-    if ispynull(_maindict.o)
-        _maindict.o = pyincref(@pycheckn ccall((@pysym :PyModule_GetDict), PyPtr, (PyPtr,), pyimport("__main__")))
+const _namespaces = Dict{Module,PyDict{String,PyObject,true}}()
+
+pynamespace(m::Module) =
+    get!(_namespaces, m) do
+        if m === Main
+            return PyDict{String,PyObject,true}(pyincref(@pycheckn ccall((@pysym :PyModule_GetDict), PyPtr, (PyPtr,), pyimport("__main__"))))
+        else
+            ns = PyDict{String,PyObject}()
+            # In Python 2, it looks like `__builtin__` (w/o 's') must
+            # exist at module namespace.  See also:
+            # http://mail.python.org/pipermail/python-dev/2001-April/014068.html
+            # https://github.com/ipython/ipython/blob/512d47340c09d184e20811ca46aaa2f862bcbafe/IPython/core/interactiveshell.py#L1295-L1299
+            if pyversion < v"3"
+                PyObject(ns).setdefault("__builtin__", builtin)
+                PyObject(ns).setdefault("__builtins__", builtin)
+            end
+            return ns
+        end
     end
-    return _maindict
-end
 
 # internal function evaluate a python string, returning PyObject, given
 # Python dictionaries of global and local variables to use in the expression,
 # and a current "file name" to use for stack traces
-function pyeval_(s::AbstractString, globals=maindict(), locals=maindict(), input_type=Py_eval_input, fname="PyCall")
-    sb = String(s) # use temp var to prevent gc before we are done with o
-    sigatomic_begin()
-    try
+function pyeval_(s::AbstractString, globals=pynamespace(Main), locals=pynamespace(Main),
+                 input_type=Py_eval_input, fname="PyCall")
+    GC.@preserve s globals locals fname disable_sigint() do
```

Comment:
Don't we need `GC.@preserve` here?

Reply:
Why?  `ccall` automatically roots its arguments.

Reply:
Ah, I see. I was caught up in the line `sb = String(s) # use temp var to prevent gc before we are done with o` in the old code.  But just removing this line and passing `s` to `ccall` is enough?  I suppose `ccall` would root the argument and then `cconvert` it?

Reply:
Wait, no.  Converted string has to be preserved for _two_ `ccall`s.  So I guess it should be

```julia
    sb = String(s)
    GC.@preserve sb disable_sigint() do
        o = PyObject(@pycheckn ccall((@pysym :Py_CompileString), PyPtr,
                                     (Cstring, Cstring, Cint),
                                     sb, fname, input_type))
        return PyObject(@pycheckn ccall((@pysym :PyEval_EvalCode),
                                         PyPtr, (PyPtr, PyPtr, PyPtr),
                                         o, globals, locals))
    end
```

Reply:
No, `sb` does not have to be preserved after the first `ccall`.

`Py_CompileString` creates its own parsed representation of the input string, and does not need the caller's string to persist after the call.

Reply:
Let me dig into when that comment was added; maybe it is an error.

Reply:
Oh, you added the comment yourself in 4d577452c80f39fdcf4a38725bebff0bf252790c … I think this is an error.

Reply:
Actually, it looks like tokentized object seems to be destroyed before `Py_CompileString` returns.  But I'm not sure the if `char *str` is passed to the code object returned by `Py_CompileString`.  The [documentation](https://docs.python.org/3/c-api/veryhigh.html#c.Py_CompileString) does not mention anything.